File: 0764. N-ary Tree Level Order Traversal/Solution03.py
/*
Submission Detail:{
    Difficulty : Medium
    Acceptance Rate : 70.51 %
    Runtime : 58 ms
    Memory Usage : 16.4 MB
    Testcase : 38 / 38 passed
    Ranking : 
        Your runtime beats 67.84 % of python submissions.
        Your memory usage beats 74.16 % of python submissions.
}
*/

"""
# Definition for a Node.
class Node(object):
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution(object):
    def levelOrder(self, root):
        listtoreturn=[]
        q= collections.deque()
        q.append(root)
        while q:
            Qlen = len(q)
            level = []
            for i in range(Qlen):
                node=q.popleft()
                if node:
                    level.append(node.val)
                    for c in node.children:
                        q.append(c)
            if level:
                listtoreturn.append(level)
        return listtoreturn

File: 0662. Maximum Width of Binary Tree/Solution03.java
/*
Submission Detail:{
    Difficulty : Medium
    Acceptance Rate : 40.49 %
    Runtime : 22 ms
    Memory Usage : 42.1 MB
    Testcase : 114 / 114 passed
    Ranking : 
        Your memory usage beats 88.97 % of java submissions.
        Your memory usage beats 00.00 % of submissions.
}
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private Integer maxWidth = 0;
    private HashMap<Integer, Integer> firstColIndexTable;

    protected void DFS(TreeNode node, Integer depth, Integer colIndex) {
        if (node == null)
            return;
        // initialize the value, for the first seen colIndex per level
        if (!firstColIndexTable.containsKey(depth)) {
            firstColIndexTable.put(depth, colIndex);
        }
        Integer firstColIndex = firstColIndexTable.get(depth);

        maxWidth = Math.max(this.maxWidth, colIndex - firstColIndex + 1);
        System.out.println(colIndex);
        // Preorder DFS. Note: it is important to put the priority on the left child
        DFS(node.left, depth + 1, 2 * colIndex);
        DFS(node.right, depth + 1, 2 * colIndex + 1);
    }

    public int widthOfBinaryTree(TreeNode root) {
        // table contains the first col_index for each level
        this.firstColIndexTable = new HashMap<Integer, Integer>();

        // start from depth = 0, and colIndex = 0
        DFS(root, 0, 0);

        return this.maxWidth;
    }
}